habitat-lab/habitat/tasks/rearrange/actions/grip_actions.py

```diff
@@ -168,3 +171,209 @@ def _grasp(self):
             self._sim.grasp_mgrs[index_grasp_manager].snap_to_marker(
                 str(attempt_snap_entity)
             )
+
+
+@registry.register_task_action
+class GazeGraspAction(MagicGraspAction):
+    def __init__(self, *args, config, sim, **kwargs):
+        super().__init__(*args, config=config, sim=sim, **kwargs)
+        self.min_dist, self.max_dist = config.gaze_distance_range
+        self.center_cone_angle_threshold = np.deg2rad(
+            config.center_cone_angle_threshold
+        )
+        self.center_cone_vector = mn.Vector3(
+            config.center_cone_vector
+        ).normalized()
+
+    @property
+    def action_space(self):
+        return spaces.Box(shape=(1,), high=1.0, low=-1.0)
+
+    @staticmethod
+    def angle_between(v1, v2):
+        cosine = np.clip(np.dot(v1, v2), -1.0, 1.0)
+        object_angle = np.arccos(cosine)
+        return object_angle
+
+    def get_camera_object_angle(self, obj_pos):
+        """Calculates angle between gripper line-of-sight and given global position."""
+
+        # Get the camera transformation
+        cam_T = self.get_camera_transform()
+
+        # Get object location in camera frame
+        cam_obj_pos = cam_T.inverted().transform_point(obj_pos).normalized()
+
+        # Get angle between (normalized) location and the vector that the camera should
+        # look at
+        obj_angle = self.angle_between(cam_obj_pos, self.center_cone_vector)
+
+        return obj_angle
+
+    def get_camera_transform(self):
+        if isinstance(self.cur_articulated_agent, SpotRobot):
+            cam_info = self.cur_articulated_agent.params.cameras[
+                "articulated_agent_arm_depth"
+            ]
+        elif isinstance(self.cur_articulated_agent, StretchRobot):
+            cam_info = self.cur_articulated_agent.params.cameras["head"]
+        else:
+            raise NotImplementedError(
+                "This robot does not have GazeGraspAction."
+            )
+
+        # Get the camera's attached link
+        link_trans = self.cur_articulated_agent.sim_obj.get_link_scene_node(
+            cam_info.attached_link_id
+        ).transformation
+        # Get the camera offset transformation
+        offset_trans = mn.Matrix4.translation(cam_info.cam_offset_pos)
+        cam_trans = link_trans @ offset_trans @ cam_info.relative_transform
+
+        return cam_trans
+
+    def get_grasp_object_mask(self, abs_obj_idx):
```

Comment:
This function can be removed, right?

Reply:
Thank you @yvsriram! Good catch. Thank you so much for this!